caffe/src/caffe/layers/molgrid_data_layer.cpp

```diff
@@ -849,16 +997,20 @@ void MolGridDataLayer<Dtype>::set_mol_info(const string& file, const vector<int>
 
 }
 
-template <typename Dtype>
-void MolGridDataLayer<Dtype>::set_grid_ex(Dtype *data, const MolGridDataLayer<Dtype>::example& ex,
-    const string& root_folder, MolGridDataLayer<Dtype>::mol_transform& transform, int pose, output_transform& peturb, bool gpu)
+template <typename Dtype, class GridMakerT>
+void BaseMolGridDataLayer<Dtype, GridMakerT>::set_grid_ex(Dtype *data, 
+    const BaseMolGridDataLayer<Dtype, GridMakerT>::example& ex,
+    const string& root_folder, BaseMolGridDataLayer<Dtype, GridMakerT>::mol_transform& transform, 
+    int pose, output_transform& peturb, bool gpu)
 {
   //set grid values for example
   //cache atom info
   //pose specifies which ligand pose to use (relevant if num_poses > 1)
   //if it is negative, use them all (but with distinct channels
   //data should be positioned at the start of the example
-  bool docache = this->layer_param_.molgrid_data_param().cache_structs();
+  const MolGridDataParameter& param = this->layer_param_.molgrid_data_param();
+  bool docache = param.cache_structs();
+  bool dream = param.dream();
```

Comment:
I do not like all the customization for dream - is unnecessary - just don't call forward on molgrid more than once per a gradient optimization run